courseworks/week1/XU_Shiqinrui/Assignment 1.md

```diff
@@ -0,0 +1,57 @@
+
+The TPU is designed to be used in large-scale data center environments and is available as a cloud service through Google Cloud Platform. This makes it easy for developers to use the TPU without needing to purchase and maintain their own hardware.
+
+Conclusion:
+Each of these specialized processors has unique features that make them suitable for different applications and workloads. The NPU is designed for low-power, embedded devices, while FPGAs are highly flexible and can be reconfigured to perform different tasks. The TPU is optimized for machine learning workloads and is available as a cloud service through Google Cloud Platform.
+
+As machine learning and artificial intelligence continue to play a larger role in our lives, specialized hardware accelerators like these are becoming increasingly important. By providing faster, more efficient ways to perform machine learning workloads, these specialized processors are helping to drive the development of new and innovative AI applications.
+### Q4
+Floating point numbers violate the associativity, commutativity, and distributivity of addition and multiplication, as well as the existence of multiplicative inverses and the distributive law, to some extent, due to the limitations of finite precision arithmetic.
+
+1.  Associativity of addition and multiplication: Due to rounding errors, the associativity of floating point addition and multiplication may not hold. This means that the result of a calculation may depend on the order in which operations are performed, which violates the associativity property.
+
+2.  Commutativity of addition and multiplication: Due to rounding errors, the commutativity of floating point addition and multiplication may not hold. This means that the result of a calculation may depend on the order of operands, which violates the commutativity property.
+
+3.  Distributivity of addition over multiplication and vice versa: Due to rounding errors, the distributive law of floating point arithmetic may not hold exactly. This means that the result of a calculation involving multiplication and addition may not be exactly equal to the result of applying the distributive law, which violates the distributivity property.
+
+4.  Existence of multiplicative inverses: Floating point arithmetic only guarantees the existence of multiplicative inverses for non-zero values, and not for zero. This means that division by zero is undefined, which violates the existence of multiplicative inverses.
+
+5.  Identity elements: Floating point arithmetic provides an additive identity element (0) and a multiplicative identity element (1), but these may not be exact values due to rounding errors. This means that the identities may not behave exactly as expected in some computations, and may not satisfy the field axioms exactly.
```

Comment:
They are exact under the defination of the floating point numbers.
You may have to consider the way they are stored in memory.